src/Functions/keyvaluepair/src/impl/state/NoEscapingValueStateHandler.h

```diff
@@ -0,0 +1,115 @@
+#pragma once
+
+#include "StateHandler.h"
+#include <unordered_set>
+
+namespace DB
+{
+
+template <QuotingStrategy QUOTING_STRATEGY>
+class NoEscapingValueStateHandler : public StateHandler
+{
+public:
+    using ElementType = std::string_view;
+
+    NoEscapingValueStateHandler(
+        char item_delimiter_,
+        std::optional<char> enclosing_character_,
+        std::unordered_set<char> special_character_allowlist_)
+        : StateHandler(enclosing_character_)
+        , item_delimiter(item_delimiter_), special_character_allowlist(special_character_allowlist_)
+    {
+    }
+
+    [[nodiscard]] NextState wait(std::string_view file, size_t pos) const
+    {
+        while (pos < file.size())
+        {
+            const auto current_character = file[pos];
+
+            if constexpr (QuotingStrategy::WithQuoting == QUOTING_STRATEGY)
+            {
+                if (current_character == enclosing_character)
+                {
+                    return {pos + 1u, State::READING_ENCLOSED_VALUE};
+                }
+            }
+            if (current_character == item_delimiter)
+            {
+                return {pos, State::READING_EMPTY_VALUE};
+            }
+            else if (isValidCharacter(current_character))
+            {
+                return {pos, State::READING_VALUE};
+            }
+            else
+            {
+                pos++;
+            }
+        }
+
+        return {pos, State::READING_EMPTY_VALUE};
+    }
+
+    [[nodiscard]] NextState read(std::string_view file, size_t pos, ElementType & value)
+    {
+        auto start_index = pos;
+
+        value = {};
+
+        while (pos < file.size())
+        {
+            const auto current_character = file[pos++];
+
+            if (current_character == item_delimiter || !isValidCharacter(current_character))
+            {
+                value = createElement(file, start_index, pos - 1);
+                return {pos, State::FLUSH_PAIR};
+            }
+        }
+
+        // TODO: do I really need the below logic?
+        // this allows empty values at the end
+        value = createElement(file, start_index, pos);
+        return {pos, State::FLUSH_PAIR};
```

Comment:
avoid unnessesary repetions, do smth like

```suggestion
        while (pos < file.size())
        {
            const auto current_character = file[pos++];

            if (current_character == item_delimiter || !isValidCharacter(current_character))
                  continue;
         }

        value = createElement(file, start_index, pos);
        return {pos, State::FLUSH_PAIR};
```